### ISSUE TITLE
XGBoost4J-Spark crashes with data having NaN value

### ISSUE THREAD

**Issue by Richardxxxxxxx** [dmlc/xgboost#4714]
Thanks for participating in the XGBoost community! We use https://discuss.xgboost.ai for any general usage questions and discussions. The issue tracker is used for actionable items such as feature proposals discussion, roadmaps, and bug tracking.  You are always welcomed to post on the forum first :)

Issues that are inactive for a period of time may get closed. We adopt this policy so that we won't lose track of actionable issues that may fall at the bottom of the pile. Feel free to reopen a new one if you feel there is an additional problem that needs attention when an old one gets closed.

For bug reports, to help the developer act on the issues, please include a description of your environment, preferably a minimum script to reproduce the problem.

For feature proposals, list clear, small actionable items so we can track the progress of the change.

**Comment by Daniel8hen** [comment]
@Richardxxxxxxx , of course it's not a solver for the issue, but just as a workaround for now, what you can do is use one of the strategies in the main documentation which I've wrote [here](https://xgboost.readthedocs.io/en/latest/jvm/xgboost4j_spark_tutorial.html#dealing-with-missing-values).

I've wrote this comment just so you'll be able to keep on with your work for now.

**Comment by Richardxxxxxxx** [comment]
> @Richardxxxxxxx , of course it's not a solver for the issue, but just as a workaround for now, what you can do is use one of the strategies in the main documentation which I've wrote [here](https://xgboost.readthedocs.io/en/latest/jvm/xgboost4j_spark_tutorial.html#dealing-with-missing-values).
> 
> I've wrote this comment just so you'll be able to keep on with your work for now.

Thanks! 
In fact, I have followed the instruction of `Keep it (using setHandleInvalid = “keep”), and set the “missing” parameter in XGBClassifier/XGBRegressor as the value representing missing. Used in (2) and (4).` 

```
val featureIndexer = new VectorIndexer()
    .setInputCol("features")
    .setOutputCol("indexedFeatures")
    .setMaxCategories(2)
    .setHandleInvalid("keep")
    .fit(data)
```
and 

`setMissing(Float.NaN)`

with the training data which looks like 

```
label feature1
1 NaN
0 0.0
0 1.0

```
It also fails to train.

**Comment by Daniel8hen** [comment]
@Richardxxxxxxx , did you try setting a dummy value which is not part of your distribution of the dataset within your model (e.g. something like 999 instead of the Float.NaN)?
Do give it a try as it works well for me and not failing.

Of course, don't forget to set the "missing" parameter in the XGBoost object.

**Comment by Daniel8hen** [comment]
the setMissing should be on the XGBoostClassifier side and not on the VectorIndexer, btw.

**Comment by Richardxxxxxxx** [comment]
@Daniel8hen , yep, replacing Float.NaN with other value like 999 and `new XGBoostClassifier().setMissing(999)` is fine.

Here is my testing code which will cause the spark job crash.

```
import org.apache.spark.ml.linalg.{Vectors,Vector}
import org.apache.spark.sql.Row
import ml.dmlc.xgboost4j.scala.spark.{XGBoostClassificationModel, XGBoostClassifier}
import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.feature._
import org.apache.spark.sql.DataFrame


val floatNaN = Vectors.dense(Float.NaN,0.0,0.0,1646.0)
val zero =     Vectors.dense(0.0,0.0,0.0,1646.0)
val one =      Vectors.dense(1.0,0.0,0.0,1646.0)



val floatNaNLabel1 = spark.sparkContext.parallelize(Seq(Row("1",floatNaN))).map { case Row(label:String, features:Vector) =>(label, features)}.toDF("label", "features")
val zeroLabel0 = spark.sparkContext.parallelize(Seq(Row("0",zero))).map { case Row(label:String, features:Vector) =>(label, features)}.toDF("label", "features")
val oneLabel1 = spark.sparkContext.parallelize(Seq(Row("1",one))).map { case Row(label:String, features:Vector) =>(label, features)}.toDF("label", "features")


val trainData = floatNaNLabel1.union(zeroLabel0).union(oneLabel1).toDF()

var xgb = new XGBoostClassifier().setLabelCol("indexedLabel").setFeaturesCol("indexedFeatures").setMinChildWeight(1.0).setEta(0.1).setMaxDepth(6).setSubsample(0.8).setColsampleBytree(0.8).setObjective("binary:logistic").setNumRound(50).setNumWorkers(1).setMissing(Float.NaN)


val labelIndexer = new StringIndexer().setInputCol("label").setOutputCol("indexedLabel").fit(trainData)
val featureIndexer = new VectorIndexer().setInputCol("features").setOutputCol("indexedFeatures").setMaxCategories(2).setHandleInvalid("keep").fit(trainData)
val pipeline = new Pipeline().setStages(Array(labelIndexer, featureIndexer, xgb))
val model = pipeline.fit(trainData)
```
But when replacing the Float.NaN with other value like:

```
val floatNaN = Vectors.dense(9.0,0.0,0.0,1646.0)
val zero =     Vectors.dense(0.0,0.0,0.0,1646.0)
val one =      Vectors.dense(1.0,0.0,0.0,1646.0)
```

There will be no error.

**Comment by Daniel8hen** [comment]
I see @Richardxxxxxxx . At least you have a workaround as I suggested in the documentation.
W.r.t the NaN issue, Yep, Someone will need to assist here more than I can.
@CodingCat can you help?
Thanks!
